### ISSUE TITLE
flaky test :  integration test  kafka_big_messages

### ISSUE THREAD

**Issue by maxshuang** [pingcap/tiflow#4731]
### Which jobs are flaking?

https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133

### Which test(s) are flaking?

 kafka_big_messages

### Jenkins logs or GitHub Actions link

```log
[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-349)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-350)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-351)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-352)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-353)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-354)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-355)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-356)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-357)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-358)[](https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9326/pipeline/133#step-1201-log-359)[2022-03-01T09:12:58.093Z] table kafka_big_messages.test not exists for 51-th check, retry later

[2022-03-01T09:13:00.155Z] table kafka_big_messages.test not exists for 52-th check, retry later

[2022-03-01T09:13:02.104Z] table kafka_big_messages.test not exists for 53-th check, retry later

[2022-03-01T09:13:04.078Z] table kafka_big_messages.test not exists for 54-th check, retry later

[2022-03-01T09:13:06.654Z] table kafka_big_messages.test not exists for 55-th check, retry later

[2022-03-01T09:13:08.098Z] table kafka_big_messages.test not exists for 56-th check, retry later

[2022-03-01T09:13:10.665Z] table kafka_big_messages.test not exists for 57-th check, retry later

[2022-03-01T09:13:12.124Z] table kafka_big_messages.test not exists for 58-th check, retry later

[2022-03-01T09:13:14.696Z] table kafka_big_messages.test not exists for 59-th check, retry later

[2022-03-01T09:13:16.628Z] table kafka_big_messages.test not exists for 60-th check, retry later

[2022-03-01T09:13:18.542Z] table kafka_big_messages.test not exists at last check

[2022-03-01T09:13:40.471Z] make: *** [integration_test_kafka] Error 1
```

### Anything else we need to know

- Does this test exist for other branches as well?
master

- Has there been a high frequency of failure lately?

**Comment by 3AceShowHand** [comment]
```
[2022/03/01 17:11:09.088 +08:00] [INFO] [main.go:223] ["Starting a new TiCDC open protocol consumer"]
[2022/03/01 17:11:09.609 +08:00] [INFO] [mysql_params.go:303] ["sink uri is configured"] [dsn="root:******@tcp(127.0.0.1:3306)/?interpolateParams=true&multiStatements=true&allow_auto_random_explicit_insert=1&readTimeout=2m&sql_mode=%22ALLOW_INVALID_DATES%2CONLY_FULL_GROUP_BY%2CNO_AUTO_VALUE_ON_ZERO%2CIGNORE_SPACE%2CNO_ENGINE_SUBSTITUTION%22&tidb_txn_mode=optimistic&time_zone=%22Asia%2FShanghai%22&timeout=2m&transaction_isolation=%22READ-COMMITTED%22&writeTimeout=2m"]
[2022/03/01 17:11:09.625 +08:00] [INFO] [mysql.go:168] ["Start mysql sink"]
[2022/03/01 17:11:09.866 +08:00] [INFO] [mysql_params.go:303] ["sink uri is configured"] [dsn="root:******@tcp(127.0.0.1:3306)/?interpolateParams=true&multiStatements=true&allow_auto_random_explicit_insert=1&readTimeout=2m&sql_mode=%22IGNORE_SPACE%2CNO_ENGINE_SUBSTITUTION%2CONLY_FULL_GROUP_BY%2CALLOW_INVALID_DATES%2CNO_AUTO_VALUE_ON_ZERO%22&tidb_txn_mode=optimistic&time_zone=%22Asia%2FShanghai%22&timeout=2m&transaction_isolation=%22READ-COMMITTED%22&writeTimeout=2m"]
[2022/03/01 17:11:09.878 +08:00] [INFO] [mysql.go:168] ["Start mysql sink"]
[2022/03/01 17:11:09.924 +08:00] [INFO] [main.go:277] ["TiCDC open protocol consumer up and running!..."]
[2022/03/01 17:11:11.086 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522018704228353]
[2022/03/01 17:11:12.085 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522020397940737]
[2022/03/01 17:11:13.096 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522020657201160]
[2022/03/01 17:11:14.221 +08:00] [INFO] [mysql.go:358] ["Exec DDL succeeded"] [sql="CREATE DATABASE `kafka_big_messages`"]
[2022/03/01 17:11:14.221 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522020670308353]
[2022/03/01 17:11:15.086 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522020827594755]
[2022/03/01 17:11:15.102 +08:00] [WARN] [main.go:495] ["receive redundant ddl job"] [ddlts=431522020827594755] [globalResolvedTs=431522020827594755]
[2022/03/01 17:11:20.385 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522021457002499]
[2022/03/01 17:11:20.586 +08:00] [ERROR] [mysql.go:722] ["execute DMLs failed"] [err="[CDC:ErrMySQLTxnError]MySQL txn error: Error 1146: Table 'kafka_big_messages.test' doesn't exist"]
[2022/03/01 17:11:20.587 +08:00] [PANIC] [main.go:272] ["Error running consumer: %v"] [error="[CDC:ErrMySQLTxnError]MySQL txn error: Error 1146: Table 'kafka_big_messages.test' doesn't exist"] [errorVerbose="[CDC:ErrMySQLTxnError]MySQL txn error: Error 1146: Table 'kafka_big_messages.test' doesn't exist\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLWithMaxRetries.func1.3\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:571\ngithub.com/pingcap/tiflow/cdc/sink.(*Statistics).RecordBatchExecution\n\tgithub.com/pingcap/tiflow/cdc/sink/statistics.go:109\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLWithMaxRetries.func1\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:558\ngithub.com/pingcap/tiflow/pkg/retry.run\n\tgithub.com/pingcap/tiflow/pkg/retry/retry_with_opt.go:54\ngithub.com/pingcap/tiflow/pkg/retry.Do\n\tgithub.com/pingcap/tiflow/pkg/retry/retry_with_opt.go:32\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLWithMaxRetries\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:550\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLs\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:721\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSinkWorker).run.func3\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql_worker.go:114\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSinkWorker).run\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql_worker.go:150\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).createSinkWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:392\nruntime.goexit\n\truntime/asm_amd64.s:1371"] [stack="main.main.func2\n\tcommand-line-arguments/main.go:272"]
```

It looks the problem happened in the consumer, table `test` is not created yet, but the `DMLs` try to flush.

**Comment by 3AceShowHand** [comment]
```
// source tidb
17:11:12.271 DROP DATABASE IF EXISTS kafka_big_messages
17:11:12.272 CREATE DATABASE kafka_big_messages
17:11:12.854 CREATE TABLE test

// cdc
17:11:14:038 CREATE DATABASE `kafka_big_messages
17:11:15.093 CREATE TABLE `test` 

// consumer
17:11:14.221 CREATE DATABASE `kafka_big_messages
.....
17:11:20.586 ["execute DMLs failed"] [err="[CDC:ErrMySQLTxnError]MySQL txn error: Error 1146: Table 'kafka_big_messages.test' doesn't exist"]

// target tidb
// create database -> show create table 
17:11:14.216 CREATE DATABASE `kafka_big_messages done

[2022/03/01 17:11:14.092 +08:00] [INFO] [session.go:3208] ["CRUCIAL OPERATION"] [conn=25] [schemaVersion=31] [cur_db=] [sql="CREATE DATABASE `kafka_big_messages`"] [user=root@%]
[2022/03/01 17:11:14.092 +08:00] [INFO] [session.go:2454] ["Try to create a new txn inside a transaction auto commit"] [conn=25] [schemaVersion=31] [txnStartTS=431522021126701057] [txnScope=global]
[2022/03/01 17:11:14.108 +08:00] [INFO] [ddl_worker.go:325] ["[ddl] add DDL jobs"] ["batch count"=1] [jobs="ID:64, Type:create schema, State:none, SchemaState:queueing, SchemaID:63, TableID:0, RowCount:0, ArgLen:1, start time: 2022-03-01 17:11:14.049 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0; "]
[2022/03/01 17:11:14.108 +08:00] [INFO] [ddl.go:615] ["[ddl] start DDL job"] [job="ID:64, Type:create schema, State:none, SchemaState:queueing, SchemaID:63, TableID:0, RowCount:0, ArgLen:1, start time: 2022-03-01 17:11:14.049 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0"] [query="CREATE DATABASE `kafka_big_messages`"]
[2022/03/01 17:11:14.112 +08:00] [INFO] [ddl_worker.go:758] ["[ddl] run DDL job"] [worker="worker 3, tp general"] [job="ID:64, Type:create schema, State:none, SchemaState:queueing, SchemaID:63, TableID:0, RowCount:0, ArgLen:0, start time: 2022-03-01 17:11:14.049 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0"]
[2022/03/01 17:11:14.156 +08:00] [INFO] [domain.go:139] ["diff load InfoSchema success"] [currentSchemaVersion=31] [neededSchemaVersion=32] ["start time"=7.26326ms] [phyTblIDs="[]"] [actionTypes="[]"]
[2022/03/01 17:11:14.191 +08:00] [INFO] [ddl_worker.go:966] ["[ddl] wait latest schema version changed"] [worker="worker 3, tp general"] [ver=32] ["take time"=53.132907ms] [job="ID:64, Type:create schema, State:done, SchemaState:public, SchemaID:63, TableID:0, RowCount:0, ArgLen:1, start time: 2022-03-01 17:11:14.049 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0"]
[2022/03/01 17:11:14.195 +08:00] [INFO] [ddl_worker.go:439] ["[ddl] finish DDL job"] [worker="worker 3, tp general"] [job="ID:64, Type:create schema, State:synced, SchemaState:public, SchemaID:63, TableID:0, RowCount:0, ArgLen:0, start time: 2022-03-01 17:11:14.049 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0"]
[2022/03/01 17:11:14.216 +08:00] [INFO] [ddl.go:679] ["[ddl] DDL job is finished"] [jobID=64]

[2022/03/01 17:11:14.216 +08:00] [INFO] [callback.go:107] ["performing DDL change, must reload"]
[2022/03/01 17:11:16.632 +08:00] [WARN] [collate.go:220] ["[ddl:1273]Unsupported collation when new collation is enabled: 'latin1_swedish_ci'"]
[2022/03/01 17:11:16.651 +08:00] [WARN] [session.go:1848] ["compile SQL failed"] [conn=27] [error="[schema:1146]Table 'kafka_big_messages.test' doesn't exist"] [SQL="show create table kafka_big_messages.test"]
[2022/03/01 17:11:16.651 +08:00] [INFO] [conn.go:1119] ["command dispatched failed"] [conn=27] [connInfo="id:27, addr:127.0.0.1:46774 status:10, collation:latin1_swedish_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="show create table kafka_big_messages.test"] [txn_mode=PESSIMISTIC] [timestamp=0] [err="[schema:1146]Table 'kafka_big_messages.test' doesn't exist"]
[2022/03/01 17:11:18.670 +08:00] [WARN] [collate.go:220] ["[ddl:1273]Unsupported collation when new collation is enabled: 'latin1_swedish_ci'"]
[2022/03/01 17:11:18.671 +08:00] [WARN] [session.go:1848] ["compile SQL failed"] [conn=29] [error="[schema:1146]Table 'kafka_big_messages.test' doesn't exist"] [SQL="show create table kafka_big_messages.test"]
```

By analyzing the time line of each component execute the DDL, it looks that the consumer does not execute the DDL.

**Comment by 3AceShowHand** [comment]
```
CREATE DATABASE kafka_big_messages -> 431522020670308353
CREATE TABLE test(id int primary key -> 431522020827594755

[2022/03/01 17:11:14.221 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522020670308353] 
[2022/03/01 17:11:15.086 +08:00] [INFO] [main.go:586] ["update globalResolvedTs"] [ts=431522020827594755]
[2022/03/01 17:11:15.102 +08:00] [WARN] [main.go:495] ["receive redundant ddl job"] [ddlts=431522020827594755] [globalResolvedTs=431522020827594755]
```
the ddl is treated as redundant, since it's committs equal to the globalResolvedTs

**Comment by 3AceShowHand** [comment]
it may cause by the case that ddl's commit ts equal to the checkpoint ts, and checkpoints was sent before the DDL sent.

add some logs first, wait for it to be reproduced.

**Comment by 3AceShowHand** [comment]
https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9637/pipeline/

**Comment by 3AceShowHand** [comment]
https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_kafka_integration_test/detail/cdc_ghpr_kafka_integration_test/9643/pipeline/140/

**Comment by 3AceShowHand** [comment]
this can be partly fixed by #4845

**Comment by 3AceShowHand** [comment]
close by $4814
